feat: add support for docker-compose.yml

misterjoshua opened the pull request:
_Issue #, if applicable:_ None

_Description of changes:_
This adds a `DockerCompose` component that can synthesize docker-compose files.

I like to use docker-compose to set up local dev environments. I think it'd be pretty neat to put this stuff under projen's management.

The interface looks like this:

```ts
new DockerCompose(project, {
  services: {
    www: {
      image: 'nginx',
      volumes: [
        DockerCompose.bindVolume('./html', '/var/www/html'),
        DockerCompose.namedVolume('files', '/var/files'),
      ],
    },
  },
});
```

...and the component is smart about reusing named volumes across services. I have a Wordpress example that does this in one of the unit tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

eladb commented on src/docker-compose.test.ts:
```diff
@@ -0,0 +1,365 @@
+    });
+
+    dc.addService('wordpress', {
+      dependsOn: ['db'],
```
Would have been cool if this was strongly coupled through a reference:

```ts
const db = dc.addService(...);
const wp = dc.addService('wordpress', {
  dependsOn: [db]
  ...
});

// or through an imperative api
wp.addDependsOn(db);
```

eladb commented on src/docker-compose.ts:
```diff
@@ -0,0 +1,427 @@
+  /**
+   * A name to add to the docker-compose.yml filename.
+   * @example 'myname' yields 'docker-compose.myname.yml'
+   * @default no name is added
```
Conventionally, if the default is not a specific value, we use "-" for the value:

```ts
@default - no name is added
```

eladb commented on src/docker-compose.ts:
```diff
@@ -0,0 +1,427 @@
+  static portMapping(
+    publishedPort: number,
+    targetPort: number,
+    protocol: DockerComposeProtocol = DockerComposeProtocol.TCP,
```
I'd move `protocol` into an options struct so it's extensible in the future. 

As a rule of thumb, no more than 2 positional arguments.

eladb commented on src/docker-compose.ts:
```diff
@@ -0,0 +1,427 @@
+   * @param name name of the service
+   * @param description a service description
+   */
+  addService(name: string, description: DockerComposeServiceDescription): void {
```
Specify access modifier:

```suggestion
  public addService(name: string, description: DockerComposeServiceDescription): void {
```

eladb reviewed (changes requested):
Great job! Looks awesome. A few minor comments.

misterjoshua commented:
Good ideas! I'll get to these changes tonight or tomorrow night. :)

misterjoshua commented:
@eladb I think I've got your review items handled. In addition to your comments, I fleshed out the imperative interfaces in general, fixed a few bugs, and added more tests.
